### PR TITLE
feat: notification system

### PR DESCRIPTION
**Pull request by sshanzel**
### Describe what this PR does
- Short and concise, bullet points can help
- Screenshots if applicable can also help

## Events

Did you introduce any new tracking events?
Don't forget to update the [Analytics Taxonomy sheet](https://docs.google.com/spreadsheets/d/18Lv7zXges9QfVX5VYL1a-Hyl0e1sQ3sLr0OK8YZWKXI/edit#gid=0)

Log the new/changed events below:

| Type   | event_name  | value |
|--------|-------------|-------|
| Change/New | event name  | extra: { ... } |

### On those affected packages:
- [ ] Have you done sanity checks in the webapp?
- [ ] Have you done sanity checks in the extension?
- [ ] Does this not break anything in companion?

### Did you test the modified components media queries?
- [ ] MobileL (420px)
- [ ] Tablet (656px)
- [ ] Laptop (1020px)

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

WT-{number} #done

### CODE REVIEW

**Review by rebelchris — CHANGES_REQUESTED** [review]
The commits should go into the intermediate right?

File: packages/shared/src/components/notifications/NotificationItem.tsx
Comment (idoshamun) [inline packages/shared/src/components/notifications/NotificationItem.tsx:32]:
```suggestion
  const hasAvatar = avatarComponents.length > 0;
```

File: packages/shared/src/components/notifications/NotificationItem.tsx
Comment (idoshamun) [inline packages/shared/src/components/notifications/NotificationItem.tsx:17]: Where do we handle the `targetUrl`?
Another suggestion is to have the root object to actually be an anchor tag that directs to `targetUrl` and we can intercept the click and change it to open modals when needed

File: packages/shared/src/components/notifications/NotificationItemAttachment.tsx
Comment (idoshamun) [inline packages/shared/src/components/notifications/NotificationItemAttachment.tsx:13]: It's not always the case. We have attachment type.
```suggestion
        alt={`Cover image of: ${title}`}
```

File: packages/shared/src/components/notifications/NotificationItem.tsx
Comment (idoshamun) [inline packages/shared/src/components/notifications/NotificationItem.tsx:31]: To filter out the nulls
```suggestion
  const avatarComponents =
    avatars?.map?.((avatar) => (
      <NotificationItemAvatar key={avatar.referenceId} {...avatar} />
    )).filter((avatar) => avatar) ?? [];
```

File: packages/webapp/pages/notifications.tsx
Comment (idoshamun) [inline packages/webapp/pages/notifications.tsx:57]: Where do we fetch the next page?

**Review by idoshamun — CHANGES_REQUESTED** [review]
Initial review comments. Just focused on the notifications page this time

File: packages/webapp/pages/notifications.tsx
Comment (sshanzel) [reply packages/webapp/pages/notifications.tsx:57]: I have created a generic component and used it in the `UpvoterList` but I actually forgot to use it here 🤣 
This should work now, I have pushed the fix.

File: packages/shared/src/components/notifications/NotificationItemAttachment.tsx
Comment (sshanzel) [reply packages/shared/src/components/notifications/NotificationItemAttachment.tsx]: I will have to replace the `image` to `preview` as the lint is not accepting it.

File: packages/shared/src/components/notifications/InAppNotification.module.css
Comment (sshanzel) [inline packages/shared/src/components/notifications/InAppNotification.module.css:12]: Convert to using tailwindcss as these ones are available.

File: packages/shared/src/components/notifications/FirstNotification.tsx
Comment (sshanzel) [inline packages/shared/src/components/notifications/FirstNotification.tsx:23]: Before merging, make the URL be pointed at our changelog for the notifications system.

Comment (idoshamun) [reply packages/shared/src/components/notifications/FirstNotification.tsx:23]: Reminder

File: packages/webapp/.eslintignore
Comment (idoshamun) [inline packages/webapp/.eslintignore:3]: Not needed any more

File: packages/webapp/pages/_app.tsx
Comment (idoshamun) [inline packages/webapp/pages/_app.tsx:143]: Fix for the issue mentioned on Slack about reseting the title

```suggestion
        titleTemplate={unreadCount ? `(${unreadText}) %s` : '%s' }
```

**Review by idoshamun — APPROVED** [review]
Looks great! I'd like Chris to review it as well before we merge.
Also another pending fix is the cropped gif that I mentioned on Slack

File: packages/shared/src/components/notifications/EnableNotification.tsx
Comment (rebelchris) [inline packages/shared/src/components/notifications/EnableNotification.tsx:64]: This is now duplicate can we mabye abstract it to one implementation somewhere?

File: packages/shared/src/components/notifications/FirstNotification.tsx
Comment (rebelchris) [reply packages/shared/src/components/notifications/FirstNotification.tsx:23]: Another reminder before merging 🥳

File: packages/shared/src/components/notifications/InAppNotification.tsx
Comment (rebelchris) [inline packages/shared/src/components/notifications/InAppNotification.tsx:26]: Nothing urgent, but thinking about this, maybe we should specify a z-index sheet at one stage.
No action required I'll note it somewhere.

File: packages/shared/src/components/notifications/InAppNotification.tsx
Comment (rebelchris) [inline packages/shared/src/components/notifications/InAppNotification.tsx:37]: I think generally we use `isExit, setIsExit` right?
Why did we decide to change it here?

File: packages/shared/src/components/notifications/InAppNotification.tsx
Comment (rebelchris) [inline packages/shared/src/components/notifications/InAppNotification.tsx:32]: In general this has a lot of timers and some useEffect, perhaps those could be abstracted in a notification timer hook?

File: packages/shared/src/components/notifications/NotificationItem.tsx
Comment (rebelchris) [inline packages/shared/src/components/notifications/NotificationItem.tsx:58]: What should this be?

File: packages/shared/src/components/notifications/Toast.spec.tsx
Comment (rebelchris) [inline packages/shared/src/components/notifications/Toast.spec.tsx:56]: 🥳

File: packages/shared/src/contexts/NotificationsContext.tsx
Comment (rebelchris) [inline packages/shared/src/contexts/NotificationsContext.tsx:46]: Is see we use this quite often actually, perhaps it's best to utilize this somewhere as one function.
If it would ever change we can easily manage it.

File: packages/shared/src/contexts/NotificationsContext.tsx
Comment (rebelchris) [inline packages/shared/src/contexts/NotificationsContext.tsx:47]: This one was very tricky to see what it actually was.
Is there not a better default state for it?

File: packages/webapp/pages/popup/notifications/enable.tsx
Comment (rebelchris) [inline packages/webapp/pages/popup/notifications/enable.tsx:83]: Why did we inline this one in particular?
(instead of hosting it?)

**Review by rebelchris — CHANGES_REQUESTED** [review]
Wow, some tremendous amounts of work has gone in this PR!
Amazing to see.

Got some remarks as a first time look over. 
Would love to hear what you think of the suggestions, overall looks very solid.

File: packages/shared/src/components/notifications/NotificationItem.tsx
Comment (sshanzel) [reply packages/shared/src/components/notifications/NotificationItem.tsx:58]: This is overridden by the `passHref` from the `Link` component.

File: packages/shared/src/components/notifications/InAppNotification.tsx
Comment (sshanzel) [reply packages/shared/src/components/notifications/InAppNotification.tsx:37]: I agree. Let's make this consistent. Will refactor this one out.

File: packages/shared/src/components/modals/PushNotificationModal.tsx
Comment (sshanzel) [reply packages/shared/src/components/modals/PushNotificationModal.tsx:26]: Yes, as we require the result of the permission given by the user.

File: packages/shared/src/components/notifications/EnableNotification.tsx
Comment (sshanzel) [reply packages/shared/src/components/notifications/EnableNotification.tsx:64]: This is a banner somewhere else. The place this is ideal to be abstracted is at `onTogglePermission` but there are other areas where it is used and has its own tracking events. This was separated intentionally to avoid duplication.

Comment (rebelchris) [reply packages/shared/src/components/notifications/EnableNotification.tsx:64]: Hmm and what if we abstract this tracking part? 
Just to avoid clutter and confusion down the line?

I think we do it for some other events already.

File: packages/shared/src/components/notifications/NotificationItem.tsx
Comment (rebelchris) [reply packages/shared/src/components/notifications/NotificationItem.tsx:58]: I think in that case we can either omit this href, right?
Or simply pass the same href just for linting purpose.

File: packages/shared/src/contexts/NotificationsContext.tsx
Comment (sshanzel) [reply packages/shared/src/contexts/NotificationsContext.tsx:47]: I'm not sure what possible default values we can use for this since this is coming from a package.

File: packages/webapp/pages/popup/notifications/enable.tsx
Comment (sshanzel) [reply packages/webapp/pages/popup/notifications/enable.tsx:83]: It was already downloadable and there is no additional layer required to render which made it suitable for SVG.

File: packages/shared/src/contexts/NotificationsContext.tsx
Comment (sshanzel) [reply packages/shared/src/contexts/NotificationsContext.tsx]: I agree. A much better procedure.

File: packages/shared/src/components/notifications/EnableNotification.tsx
Comment (sshanzel) [reply packages/shared/src/components/notifications/EnableNotification.tsx:64]: Sure thing. Let me create a generic function for this 🚢

File: packages/shared/src/components/notifications/NotificationItem.tsx
Comment (sshanzel) [reply packages/shared/src/components/notifications/NotificationItem.tsx:58]: Hmm. Why I didn't think of just passing the same `targetUrl` as you mentioned in the first place 🤣

File: packages/shared/src/contexts/NotificationsContext.tsx
Comment (rebelchris) [reply packages/shared/src/contexts/NotificationsContext.tsx:48]: Hmm alright, leave it as is, nothing urgent 🥳

File: packages/shared/src/lib/links.ts
Comment (rebelchris) [inline packages/shared/src/lib/links.ts:32]: We already keep track of these in `lib/constants.ts` please keep all links there.

File: packages/webapp/pages/popup/notifications/enable.tsx
Comment (rebelchris) [inline packages/webapp/pages/popup/notifications/enable.tsx:22]: We should probably generalize this app wide, but no need to fix it for now.
Perhaps you can log it as a ticket in the backlog.

File: packages/webapp/pages/popup/notifications/enable.tsx
Comment (rebelchris) [inline packages/webapp/pages/popup/notifications/enable.tsx:62]: Is the capitalisation correct here?

**Review by rebelchris — CHANGES_REQUESTED** [review]
Wohoo almost there, just a couple of remarks.

File: packages/shared/src/lib/links.ts
Comment (sshanzel) [reply packages/shared/src/lib/links.ts:32]: The constants file was actually looking a bit messy with all the links around and other constants. So I thought to store them in an object like this and migrate the others in the future. WDYT?

File: packages/webapp/pages/popup/notifications/enable.tsx
Comment (sshanzel) [reply packages/webapp/pages/popup/notifications/enable.tsx:62]: Should be accurate from the discussion here: https://dailydotdev.slack.com/archives/C03RA0WEGEQ/p1672292526819369?thread_ts=1672057398.833569&cid=C03RA0WEGEQ

File: packages/shared/src/components/notifications/EnableNotification.tsx
Comment (sshanzel) [reply packages/shared/src/components/notifications/EnableNotification.tsx:134]: Should be accurate from the design spec 💯 

image

File: packages/shared/src/components/notifications/InAppNotificationItem.tsx
Comment (sshanzel) [reply packages/shared/src/components/notifications/InAppNotificationItem.tsx:48]: Good point. Let me refactor this quickly! 🚢

File: packages/shared/src/lib/links.ts
Comment (rebelchris) [reply packages/shared/src/lib/links.ts:32]: Rather raise it as a separate issue, I don't want to have links all over the place.
It will become messy real quick.

So either we fix all links to this format or skip it for now.

File: packages/shared/src/components/notifications/InAppNotificationItem.tsx
Comment (sshanzel) [reply packages/shared/src/components/notifications/InAppNotificationItem.tsx:48]: Oh, sorry. I checked it is actually good to use this approach when handling `null` values. JS crashes when you pass a null value.

File: packages/shared/src/lib/links.ts
Comment (sshanzel) [reply packages/shared/src/lib/links.ts:32]: Good point. Let's keep everything in the constants for now and I'll create a task in the backlog to clean it up.

File: packages/webapp/pages/popup/notifications/enable.tsx
Comment (sshanzel) [reply packages/webapp/pages/popup/notifications/enable.tsx:22]: I agree. We will need to come up with a very intuitive naming based on our guidelines. Let me log it quickly.

File: packages/shared/src/components/notifications/InAppNotificationItem.tsx
Comment (sshanzel) [reply packages/shared/src/components/notifications/InAppNotificationItem.tsx:42]: I think I can do this now while others are still testing out the system.

Comment (sshanzel) [reply packages/shared/src/components/notifications/InAppNotificationItem.tsx]: What do you think of the changes below?
https://github.com/dailydotdev/apps/pull/1413/commits/d2bc39cddc005c3ab492550281b639dead18f37c

Comment (rebelchris) [reply packages/shared/src/components/notifications/InAppNotificationItem.tsx]: Wohoo! Amazing work 🥳

**Comment by sshanzel** [comment]
@rebelchris feel free to approve if you feel the implementations are now up to our standard 🥳 🥳 🥳

**Review by rebelchris — APPROVED** [review]
Let's go 🥳

**Pull request merged**
